tsdb/locker.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tsdb
+
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+)
+
+const (
+	lockfileDisabled       = -1
+	lockfileReplaced       = 0
+	lockfileCreatedCleanly = 1
+)
+
+type Locker struct {
+	logger log.Logger
+
+	createdCleanly prometheus.Gauge
+
+	releaser fileutil.Releaser
+	path     string
+}
+
+// NewLocker creates a locker that can obtain an exclusive lock on dir.
+func NewLocker(dir, subsystem string, l log.Logger, r prometheus.Registerer) (*Locker, error) {
+	lock := &Locker{
+		logger: l,
+		createdCleanly: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: "prometheus",
+			Subsystem: subsystem,
+			Name:      "clean_start",
```

Comment:
We usually use the full metric name. So to be consistent and easily readable, can you please have it has `Name: fmt.Sprintf("prometheus_%s_clean_start", subsystem)`? (Still having the subsystem variable because now this will be a library and it can be used to avoid metric name clashes)